src/bloaty.cc

```diff
@@ -516,35 +516,61 @@ class Rollup {
 
 void Rollup::CreateRows(RollupRow* row, const Rollup* base,
                         const Options& options, bool is_toplevel) const {
-  row->vmcapacity = vm_capacity_;
-  row->filecapacity = file_capacity_;
+  row->capacity.vm = vm_capacity_;
+  row->capacity.file = file_capacity_;
   row->capacity_origin = capacity_origin_;
 
   // If a row has a capacity, percentages are calculated relative to it.
-  if (row->vmcapacity > 0) {
-    row->vmpercent = Percent(row->vmsize, row->vmcapacity);
+  if (row->capacity.vm > 0) {
+    row->vmpercent = Percent(row->size.vm, row->capacity.vm);
   }
-  if (row->filecapacity > 0) {
-    row->filepercent = Percent(row->filesize, row->filecapacity);
+  if (row->capacity.file > 0) {
+    row->filepercent = Percent(row->size.file, row->capacity.file);
   }
 
   if (base) {
     // For a diff, the percentage is a comparison against the previous size of
     // the same label at the same level.
-    if (row->vmcapacity == 0) {
+    if (row->capacity.vm == 0) {
       row->vmpercent = Percent(vm_total_, base->vm_total_);
     }
-    if (row->filecapacity == 0) {
+    if (row->capacity.file == 0) {
       row->filepercent = Percent(file_total_, base->file_total_);
     }
   }
 
   for (const auto& value : children_) {
-    if (value.second->vm_total_ != 0 || value.second->file_total_ != 0) {
+    int64_t vm_total = value.second->vm_total_;
+    int64_t file_total = value.second->file_total_;
+    Rollup* base_child = nullptr;
+
+    if (base) {
+      // Perform the diff mode subtraction.
+      auto it = base->children_.find(value.first);
+      if (it != base->children_.end()) {
+        base_child = it->second.get();
+        vm_total -= base_child->vm_total_;
+        file_total -= base_child->file_total_;
+      }
+    }
+
+    if (vm_total != 0 || file_total != 0) {
       row->sorted_children.emplace_back(value.first);
       RollupRow& child_row = row->sorted_children.back();
-      child_row.vmsize = value.second->vm_total_;
-      child_row.filesize = value.second->file_total_;
+      child_row.size.vm = vm_total;
+      child_row.size.file = file_total;
+
+      if (base_child) {
```

Comment:
Consider peppering in some more comments-- it's not totally obvious what is going on here.